### PR TITLE
validations, model specs, and n+1 problems

### PR DESCRIPTION
**Pull request by victorokoroji**
# Validations, Model specs, and n+1 problems
_In this milestone,_ 

## I added validations for the following requirements;
- For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
- For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
 - I added unit specs for all models' methods and validations.
 - I added flash messages in the create actions in all your controllers.
- I make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframe

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -20,10 +20,10 @@ def create
     @post.likes_counter = 0
 
     if @post.save
-      flash[:success] = 'Post saved successfully'
+      flash[:alert] = 'Post saved successfully'
```
Comment (Mre55) [inline app/controllers/posts_controller.rb:23]: Here in your `PostController` class inside the `index` method when you fetch the `post` at line 5 N+1 problem has to be solved. 
image

Consider trying something like this:

```
@posts = @user.posts.includes(:comments)
```
That way all comments for that particular post will be fetched properly. Make sure after solving this problem your app is working as expected.

**Review by Mre55 — CHANGES_REQUESTED** [review]
**Hi @vickymarz,**

You have done a great job so far! 💪 👏 
You still need to work on a few issues before moving on to the next project, but you're almost there!

## To highlight 🎯

- The Pull Request is following the GitHub Flow ✅ 
- All test are passing ✅ 
- The PR title and description are professional and descriptive ✅ 
- The commit messages are meaningful ✅ 
- The README is well documented ✅ 
- Linters are working properly ✅ 

## Required Changes ♻️🛠️ 
_Check the comments under the review._




## Optional suggestions 🟡 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Gambit142 — APPROVED** [review]
Hi @vickymarz 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

### Great work implementing the suggestion made by the previous reviewer. You did great. 👍 

### Optional Suggestion
- In your _**app/views/users/_user.html.erb**_ view file, kindly change the `image_tag` element you are using to HTML `img` tag. If you implement this suggestion, your code will look like this 
```ruby on rails
<img src="<%= user.photo %> alt="Alt Name"">
```

The reason why I asked you to change it is that `image_tag` is an asset pipeline helper and it will look in your asset folder for the URL of the photo. You can use this tag if you are uploading images from your local machine or if you stored an image in the asset folder. 👍 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by victorokoroji** [comment]
Thank you for the suggestion, I wills surely implement it.

**Comment by Gambit142** [comment]
> Thank you for the suggestion, I wills surely implement it.

@vickymarz, nice work man

**Pull request merged**